src/compiler/scala/tools/nsc/typechecker/Namers.scala

```diff
@@ -1413,13 +1413,25 @@ trait Namers extends MethodSynthesis {
               case mt => (Nil, mt)
             }
 
-          @tailrec @inline def applyFully(tp: Type, paramss: List[List[Symbol]]): Type =
-            if (paramss.isEmpty) tp match {
-              case NullaryMethodType(rtpe) => rtpe
-              case MethodType(Nil, rtpe)   => rtpe
-              case tp                      => tp
-            }
-            else applyFully(tp.resultType(paramss.head.map(_.tpe)), paramss.tail)
+           @tailrec @inline def applyFully(tp: Type, paramss: List[List[Symbol]]): Type =
+             paramss match {
+               case ps :: rest =>
+                 val resTp =
+                   tp match {
+                     case MethodType(qs, SingleType(NoPrefix, sym)) if qs.contains(sym) =>
+                       val i = qs.indexWhere(_ == sym)
+                       singleType(NoPrefix, ps(i))
```

Comment:
This is the relevant new case. Is it sufficient? It seems to handle `f(p): p.type` but not all dependencies.